File: VirtualRide.cpp
/*
MIT License

Copyright (c) 2021-2022 jp-96

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

#include "pxt.h"
#include "lib/BLEFitnessMachineService.h"

namespace virtualride {
    
    BLEFitnessMachineService* _pService = NULL;
    int _lastResistanceLevel = 1;

    //%
    void advertise() {
        if (NULL!=_pService) return;
        _pService = new BLEFitnessMachineService();
    }

    //%
    void setResistanceLevel(int resistanceLevel) {
        _lastResistanceLevel = resistanceLevel;
    }
    
    //%
    void notifyIndoorBikeData(int speed100, int cadence2, int power) {
        if (NULL==_pService) return;
        _pService->notifyIndoorBikeData(speed100, cadence2, _lastResistanceLevel, power);
    }

    // Coefficient of Cadence and Speed
    static const uint64_t K_CRANK_CADENCE =  120000000;
    static const uint64_t K_CRANK_SPEED   = 1800000000;

    // https://diary.cyclekikou.net/archives/15876
    static const double K_POWER = 0.8 * (70 * 9.80665) / (360 * 0.95 * 100); // weight(70kg)
    static const double K_INCLINE_A = 0.9; // Incline(%) - a
    static const double K_INCLINE_B = 0.6; // Incline(%) - b

    //%
    void calcAndNotifyIndoorBikeData(int crankIntervalTime) {
        int speed100;
        int cadence2;
        int power;
        if (crankIntervalTime<=0)
        {
            cadence2 = 0;
            speed100 = 0;
            power = 0;
        }
        else
        {
            cadence2  = (int)( K_CRANK_CADENCE / crankIntervalTime );
            speed100 = (int)( K_CRANK_SPEED   / crankIntervalTime );
            // https://diary.cyclekikou.net/archives/15876
            power = (int)( (double)(speed100) * (K_INCLINE_A * ((double)_lastResistanceLevel + K_INCLINE_B) * K_POWER) );
        }
        notifyIndoorBikeData(speed100, cadence2, power);
    }

    /**
     * The Fitness Machine Control Point is used by a Client to control certain behaviors of the Server. 
     */
    enum virtualrideFMCP {
        //% block="Set Target Resistance Level Procedure"
        SET_TARGET_RESISTANCE_LEVEL = 0x04,
        //% block="Set Indoor Bike Simulation Parameters Procedure"
        SET_INDOOR_BIKE_SIMULATION = 0x11
    };

    //%
    void onFitnessMachineControlPoint(virtualrideFMCP op, Action body) {
        switch (op)
        {
            case virtualrideFMCP::SET_TARGET_RESISTANCE_LEVEL:
                registerWithDal(CUSTOM_EVENT_ID_VIRTUAL_RIDE, VIRTUAL_RIDE_EVT_SET_TARGET_RESISTANCE_LEVEL, body);
                break;
            case virtualrideFMCP::SET_INDOOR_BIKE_SIMULATION:
                registerWithDal(CUSTOM_EVENT_ID_VIRTUAL_RIDE, VIRTUAL_RIDE_EVT_SET_INDOOR_BIKE_SIMULATION, body);
                break;
            default:
                break;
        }
    }

    //%
    int getTargetResistanceLevel10() {
        if (NULL==_pService) return VAL_MINIMUM_RESISTANCE_LEVEL;
        return _pService->getTargetResistanceLevel10();
    }

    //%
    int getGrade100() {
        if (NULL==_pService) return 0;
        return _pService->getGrade100();
    }

}

File: VirtualRide.ts
/*
MIT License

Copyright (c) 2021 jp-96

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

/**
* Use this file to define custom functions and blocks.
* Read more at https://makecode.microbit.org/blocks/custom
*/

// enums: see VirtualRide.cpp, and also enums.d.ts

/**
 * Custom blocks  
 * icon: a Unicode identifier for an icon from the Font Awesome icon set.  
 *       http://fontawesome.io/icons  
 *       icon="\uf206" https://fontawesome.com/v5.15/icons/bicycle  
 */
//% weight=100 color=#fd7e14 icon="\uf206"
//% block="Virtual Ride"
namespace virtualride {

    /**
     * Start and advertise the Fitness Machine Service.
     */
    //% blockId=virtualride_advertise
    //% block="Start the Fitness Machine Service"
    //% shim=virtualride::advertise
    export function advertise(): void {
        return;
    }

    /**
     * Set resistanceLevel 
     * @param resistanceLevel Resistance Level (1 - 8)
     */
    //% blockId=virtualride_set_resistance_level
    //% block="Set Resistance Level %resistanceLevel"
    //% resistanceLevel.min=1 resistanceLevel.max=8 resistanceLevel.defl=1
    //% shim=virtualride::setResistanceLevel
    export function setResistanceLevel(resistanceLevel: number): void {
        return;
    }
    
    /**
     * The Indoor Bike Data characteristic is used to send training-related data to the Client from an indoor bike (Server). 
     * @param speed100 Instantaneous Speed (uint16), Kilometer per hour with a resolution of 0.01, eg: 42.20km/h -> 4220. 
     * @param cadence2 Instantaneous Cadence (uint16), 1/minute with a resolution of 0.5, eg: 60.5rpm -> 121.
     * @param power Instantaneous Power (sint16), Watts with a resolution of 1, eg:200watt -> 200
     */
    //% blockId=virtualride_notify_indoor_bike_data
    //% block="Notify Indoor Bike Data speed100:=%speed100 cadence2:=%cadence2 power:=%power"
    //% speed100.min=0 speed100.max=9900 speed100.defl=3500
    //% cadence2.min=0 cadence2.max=1000 cadence2.defl=240
    //% power.min=0 power.max=500 power.defl=220
    //% shim=virtualride::notifyIndoorBikeData
    export function notifyIndoorBikeData(speed100: number, cadence2: number, power: number): void {
        return;
    }

    /**
     * Calc speed100, cadence2, power by crankIntervalTime and resistanceLevel, and then call notifyIndoorBikeData. 
     * @param crankIntervalTime Crank Interval Time (microsecond)
     */
    //% blockId=virtualride_calc_and_notify_indoor_bike_data
    //% block="Calc And Notify Indoor Bike Data CrankIntervalTime:=%crankIntervalTime"
    //% crankIntervalTime.min=0 crankIntervalTime.max=120000000 crankIntervalTime.defl=1000000
    //% shim=virtualride::calcAndNotifyIndoorBikeData
    export function calcAndNotifyIndoorBikeData(crankIntervalTime: number): void {
        return;
    }

    /**
     * Register code to run when requested procedure through the Fitness Machine Control Point characteristic.
     * @param op The Fitness Machine Control Point Procedure.
     * @param body Code to run when the Fitness Machine Control Point characteristic requested.
     */
    //% blockId=virtualride_on_fitness_machine_control_point
    //% block="On %op through Fitness Machine Control Point"
    //% shim=virtualride::onFitnessMachineControlPoint
    export function onFitnessMachineControlPoint(op: virtualrideFMCP, body: () => void): void {
        return;
    }

    /**
     * Get Target Resistance Level through the `Set Target Resistance Level Procedure`.
     * Target Resistance Level, UINT8, Unitless with a resolution of 0.1.
     */
    //% blockId=virtualride_get_target_resistance_level_10
    //% block="Get Target Resistance Level"
    //% shim=virtualride::getTargetResistanceLevel10
    export function getTargetResistanceLevel10(): number {
        return 10;
    }

    /**
     * Get Grade through the `Set Indoor Bike Simulation Parameters Procedure`.
     * Grade, SINT16, Percentage with a resolution of 0.01.
     */
    //% blockId=virtualride_get_grade_100
    //% block="Get Simulation Grade"
    //% shim=virtualride::getGrade100
    export function getGrade100(): number {
        return 0;
    }

}
